tools/distpartitioning/dist_lookup.py

```diff
@@ -0,0 +1,289 @@
+        '''
+
+        #find the process where global_nid --> partition-id(owner) is stored. 
+        ntype_ids, type_nids = self.id_map(global_nids)
+        ntype_ids, type_nids = ntype_ids.numpy(), type_nids.numpy()
+        assert len(ntype_ids) == len(global_nids)
+
+        #For each node-type, the per-type-node-id <-> partition-id mappings are
+        #stored as contiguous chunks by this lookup service. 
+        #The no. of these mappings stored by each process, in the lookup service, are
+        #equally split among all the processes in the lookup service, deterministically
+        service_owners = self.ntype_count[ntype_ids]
+        service_owners = np.ceil(service_owners/self.world_size).astype(np.int64)
+        service_owners = np.floor_divide(type_nids, service_owners).astype(np.int64)
+
+        #now `service_owners` is a list of ranks (process-ids) which own the corresponding
+        #global-nid <-> partition-id mapping.
+
+        #Split the input global_nids into a list of lists where each list will be 
+        #sent to the respective rank/process
+        #We also need to store the indices, in the indices_list, so that we can re-order
+        #the final result (partition-ids) in the same order as the global-nids (function argument)
+        send_list = []
+        indices_list = []
+        for idx in range(self.world_size):
+            cond = service_owners == idx
+            idxes = np.where(cond)
+            ll = global_nids[idxes[0]]
+            send_list.append(torch.from_numpy(ll))
+            indices_list.append(idxes[0])
+        assert len(np.concatenate(indices_list)) == len(global_nids)
+        assert np.all(np.sort(np.concatenate(indices_list)) == np.arange(len(global_nids)))
+
+        #send the request to everyone else.
+        #As a result of this operation, the current process also receives a list of lists
+        #from all the other processes. 
+        #These lists are global-node-ids whose global-node-ids <-> partition-id mappings 
+        #are owned/stored by the current process
+        owner_req_list = alltoallv_cpu(self.rank, self.world_size, send_list)
+
+        #create the response list here for each of the request list received in the previous
+        #step. Populate the respective partition-ids in this response lists appropriately
+        out_list = []
+        for idx in range(self.world_size):
+            if owner_req_list[idx] is None:
+                out_list.append(torch.empty((0,), dtype=torch.int64))
+                continue
+            #get the node_type_ids and per_type_nids for the incoming global_nids
+            ntype_ids, type_nids = self.id_map(owner_req_list[idx].numpy())
+            nypte_ids, type_nids = ntype_ids.numpy(), type_nids.numpy()
+
+            #lists to store partition-ids for the incoming global-nids
+            type_id_lookups = []
+            local_order_idx = []
+            for tid in range(len(self.partid_list)):
```

Comment:
this loop iterates over all types node types. i think you should write comments like that.
also, the for loop iterates over `self.partid_list`. this is very difficult for people to understand it iterates over node types.

Reply:
Added comments to indicate the intention